src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,391 @@
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client: BaseClient):
+        self.client = client
+        self.root_node: Union[None, Container] = None
+        self._desc_nodes: Dict[str, Container] = {}  # references to descriptor containers by their uid's
+        self._sres_nodes: Dict[str, BaseClient] = {}
+        self._sres_cache: Dict[str, StreamResource] = {}
+        self._handlers: Dict[str, StreamHandlerBase] = {}
+
+    def _ensure_sres_backcompat(self, sres: StreamResource) -> StreamResource:
+        """Kept for back-compatibility with old StreamResource schema from event_model<1.20.0
+
+        Will make changes to and return the _same_instance_ of StreamRsource dictionary.
+        """
+
+        if ("mimetype" not in sres.keys()) and ("spec" not in sres.keys()):
+            raise RuntimeError("StreamResource document is missing a mimetype or spec")
+        else:
+            sres["mimetype"] = sres.get("mimetype", MIMETYPE_LOOKUP[sres["spec"]])
+        if "parameters" not in sres.keys():
+            sres["parameters"] = sres.pop("resource_kwargs", {})
+        if "uri" not in sres.keys():
+            file_path = sres.pop("root").strip("/") + "/" + sres.pop("resource_path").strip("/")
+            sres["uri"] = "file://localhost/" + file_path
+
+        return sres
+
+    def start(self, doc):
+        self.root_node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        if self.root_node is None:
+            raise RuntimeError("RunWriter is properly initialized: no Start document has been recorded.")
+
+        doc = dict(doc)
+        doc.pop("start", None)
+        metadata = {"stop": doc, **dict(self.root_node.metadata)}
+        self.root_node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc: EventDescriptor):
+        if self.root_node is None:
+            raise RuntimeError("RunWriter is properly initialized: no Start document has been recorded.")
+
+        desc_name = doc["name"]
+        metadata = dict(doc)
+
+        # Remove variable fields of the metadata and encapsulate them into sub-dictionaries with uids as the keys
```

Comment:
Just noting that I think we may want to take a critical look at how the information from the Event Descriptor is encoded in `metadata` at some point, but that may best be part of a larger discussion about the whole structure here. I want to say that I think this is a good approach, and probably not the final approach, and that it's too early to know what the right final approach is, so we should move ahead.